kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -306,48 +318,93 @@ protected final Optional<List<FileStatus>> listDeltaAndCheckpointFiles(
             ));
         logger.debug("startVersion: {}, versionToLoad: {}", startVersion, versionToLoad);
 
-        return listFromOrNone(engine,
-            startVersion).map(fileStatusesIter -> {
-                final List<FileStatus> output = new ArrayList<>();
-
-                while (fileStatusesIter.hasNext()) {
-                    final FileStatus fileStatus = fileStatusesIter.next();
+        List<Commit> unbackfilledCommits = commitCoordinatorClientHandlerOpt
+                .map(commitCoordinatorClientHandler -> commitCoordinatorClientHandler
+                        .getCommits(
+                                logPath.toString(),
+                                coordinatedCommitsTableConf,
+                                startVersion,
+                                versionToLoad.orElse(null))
+                        .getCommits())
+                .orElse(Collections.emptyList());
+
+
+        AtomicLong maxDeltaVersionSeen = new AtomicLong(startVersion - 1);
+        Optional<List<FileStatus>> resultFromFsListingOpt = listFromOrNone(engine,
+                startVersion).map(fileStatusesIter -> {
+                    final List<FileStatus> output = new ArrayList<>();
+
+                    while (fileStatusesIter.hasNext()) {
+                        final FileStatus fileStatus = fileStatusesIter.next();
+                        final String fileName = getName(fileStatus.getPath());
```

Comment:
Can you try undoing this change? I feel that the diff will become less complex with that.